### PR TITLE
fix: serialize secrets to json.

### PR DESCRIPTION
**Pull request by Atheuz**
## Change Summary

This change adds JSON serialization of secret types, `SecretStr` and `SecretBytes`. Also modifies the __str__ functions of the same.
It does not support serialization of the actual serialized values, which it maybe should.

## Related issue number

#462

## Checklist

* [x] Unit tests for the changes exist
* [x] Tests pass on CI and coverage remains at 100%
* [x] Documentation reflects the changes where applicable

### CODE REVIEW

**Comment by Atheuz** [comment]
I don't know if this is a good idea or not. I think @samuelcolvin would prefer not to do something like this.

File: pydantic/types.py [hunk 1]
```diff
@@ -590,7 +590,7 @@ def __repr__(self) -> str:
         return "SecretStr('**********')" if self._secret_value else "SecretStr('')"
 
     def __str__(self) -> str:
-        return repr(self)
+        return '**********' if self._secret_value else ''
```
Comment (samuelcolvin) [inline pydantic/types.py]: please don't modify these, they were fine as they are.

File: docs/examples/ex_secret_types.py [hunk 2]
```diff
@@ -6,13 +6,27 @@ class SimpleModel(BaseModel):
     password: SecretStr
     password_bytes: SecretBytes
 
-print(SimpleModel(password='IAmSensitive', password_bytes=b'IAmSensitiveBytes'))
+sm1 = SimpleModel(password='IAmSensitive', password_bytes=b'IAmSensitiveBytes')
+print(sm)
 # > SimpleModel password=SecretStr('**********') password_bytes=SecretBytes(b'**********')
+print(sm.password)
```
Comment (samuelcolvin) [inline docs/examples/ex_secret_types.py]: all this is unnecessary, it was fine as it was.

File: pydantic/json.py [hunk 3]
```diff
@@ -20,6 +22,8 @@ def isoformat(o: Union[datetime.date, datetime.time]) -> str:
     IPv6Interface: str,
     IPv4Network: str,
     IPv6Network: str,
+    SecretStr: str,
+    SecretBytes: str,
```
Comment (samuelcolvin) [inline pydantic/json.py]: This (plus tests) is the only change required for this fix.

**Comment by Atheuz** [comment]
@samuelcolvin 

The reason I made changes to the __str__ function for both is that if we keep it as is, then we end up with these kinds of things as JSON output:

```
>>> import pydantic
>>> class Settings(pydantic.BaseModel):
...     passw: pydantic.SecretStr
...
>>> conf = Settings(passw="1234")
>>> conf
<Settings passw=SecretStr('**********')>
>>> conf.dict()
{'passw': SecretStr('**********')}
>>> conf.json()
'{"passw": "SecretStr(\'**********\')"}'
```

Where as with the modified __str__ we would get something like this instead:

```
>>> import pydantic
>>> class Settings(pydantic.BaseModel):
...     passw: pydantic.SecretStr
...
>>> conf = Settings(passw="1234")
>>> conf
<Settings passw=SecretStr('**********')>
>>> conf.dict()
{'passw': SecretStr('**********')}
>>> conf.json()
'{"passw": "**********"}'
```

Please let me know what you think.

**Comment by samuelcolvin** [comment]
Ok, let's add a `display()` method to both with returns either `None` or `**********` and use that in JSON, then keep `str` and `repr` the same.

**Comment by samuelcolvin** [comment]
That will then need one line the docs/example to explain/show it.

**Comment by Atheuz** [comment]
@samuelcolvin is that something like what you were thinking?

**Comment by samuelcolvin** [comment]
Yes exactly.

File: pydantic/types.py [hunk 4]
```diff
@@ -592,6 +592,9 @@ def __repr__(self) -> str:
     def __str__(self) -> str:
         return repr(self)
 
+    def display(self) -> str:
+        return "**********" if self._secret_value else ''
```
Comment (samuelcolvin) [inline pydantic/types.py]: single quotes always please.

File: docs/examples/ex_secret_types.py [hunk 5]
```diff
@@ -5,7 +5,7 @@
 class SimpleModel(BaseModel):
     password: SecretStr
     password_bytes: SecretBytes
-
+
```
Comment (samuelcolvin) [inline docs/examples/ex_secret_types.py]: should lint these files, but please remove trailing spaces.

**Comment by Atheuz** [comment]
@samuelcolvin I've been thinking about how to deal with a situation such as this:

```
from datetime import datetime
from typing import List, Any, Union, Dict
from pydantic import BaseSettings, BaseModel, SecretStr, validator
import os

class Test:
    def __init__(self, config):
        self.config = config
        self.uri = f"http://{config['USERNAME']}:{config['PASSWORD']}@{config['HOST']}"  # Won't work!
    

class NestedSettings(BaseModel):
    HOST: str
    PASSWORD: SecretStr
    USERNAME: SecretStr
    

class Settings(BaseModel):
    USERNAME: SecretStr = SecretStr("")
    PASSWORD: SecretStr = SecretStr("")
    HOST: str = "postgres.github.io"
    PORT: int = 5432
    DB_NAME: str = "testdb"
    VARS: NestedSettings
    
    DB_URI: SecretStr = SecretStr("")
    
    
    @validator("DB_URI", pre=True, whole=True, always=True)
    def set_db_uri(cls, v, values, **kwargs):
        """Set DB_URI."""
        username = values["USERNAME"].get_secret_value()
        password = values["PASSWORD"].get_secret_value()
        host = values["HOST"]
        port = values["PORT"]
        db_name = values["DB_NAME"]
        result = f"postgresql://{username}:{password}@{host}:{port}/{db_name}"
        return result
        
    
def main():
    external_data = {"USERNAME": "abc", "PASSWORD": "def", "VARS": {"HOST": "www.google.com", "USERNAME": "user1", "PASSWORD":"passw"}}
    conf = Settings(**external_data)
    print(conf.dict())
    testobj = Test(conf.dict()["VARS"])
    print(testobj.uri)
    
if __name__ == '__main__':
    main()
```

Which would result in:

```
{'VARS': {'HOST': 'www.google.com', 'PASSWORD': SecretStr('**********'), 'USERNAME': SecretStr('**********')}, 'USERNAME': SecretStr('**********'), 'PASSWORD': SecretStr('**********'), 'HOST': 'postgres.github.io', 'PORT': 5432, 'DB_NAME': 'testdb', 'DB_URI': SecretStr('**********')}
http://SecretStr('**********'):SecretStr('**********')@www.google.com
```

If you need the inner 'object' to be provided for the configuration of another item, you have to kind of force it to fit by modifying the values in-place to be what you want. Like for instance if you want to configure object `Test` and it takes a Dict[str, str], but the object that you have is more like Dict[str, Union[str, SecretStr]] then that won't be valid and you'd have to unsecret the secret items(recursively turn everything into `str`) before you can proceed.

Which is not especially useful, ideally we would be able to unsecret the entire dict recursively and then pass it on to classes or functions that need those items to be plain-text.

Do you have any thoughts on this?

**Comment by samuelcolvin** [comment]
See [dangerouslySetInnerHTML](https://reactjs.org/docs/dom-elements.html#dangerouslysetinnerhtml) in react.

See [hazmat](https://cryptography.io/en/latest/#layout) in cryptography.

The whole point is to make it somewhat involved to get the raw value.

If there's any point whatsoever in include `SecretStr` in pydantic, then we have to make accessing the raw value only possible with `get_secret_value()` or an internal value.

If people don't need it to be secret, they don't have to use it.

In fact in your example above, it's still not exactly complicated to generate the uri:

``py
f'postgresql://{username.get_secret_value()}:{password.get_secret_value()}@{host}:{port}/{db_name}'
```

**Comment by Atheuz** [comment]
@samuelcolvin In the postgresql example, that would expose secrets if the conf object is printed out. 
I think I just need to adjust how I would be using these and use the raw values ocne the config has been loaded, in places where those values will not be exposed.

Anyway, that's not really related to this PR, just a couple of thoughts I was having. 
Let me know if you want me to make further changes.

**Comment by samuelcolvin** [comment]
> that would expose secrets if the conf object is printed out.

Yes, obviously!

If you use the secret values then they may be visible in the context in which you use them, there's no way around that.

The only point in `SecretStr` is to avoid it be printed in exceptions or logging (facebook!) until they're used.

**Pull request merged**